IA-632 leaflet

beygorghor opened the pull request:
Upgrade leaflet (1.9.3) and react-leaflet(3.2.5).

Related JIRA tickets : IA-632

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [x] New translations have been added or updated if new strings have been introduced in the frontend
- [ ] My migrations file are included
- [ ] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Changes

- Remove map reducer 
- upgrade every maps to support new versions of packages
- rewrite of custom zoom bar
- provide some docs for frontend and especially for maps
- use only one tile layer component

## How to test

Run `npm ci` first.

Go trough every maps in Iaso and check that everything is ok:
- GeoJsonMapComponent => visible in org unit history for an org unit with a shape
- MarkerMapComponent => visible in instance details with a geo location
- AssignmentsMap => planning
- ListMap => entities /beneficiaries
- InstancesMap => instances list
- OrgUnitsMap => org unit list
- OrgUnitMap => org unit detail
- OrgUnitChildrenMap => registry
- CalendarMap => Polio/calendar

## Notes

Bluesquare component as been upgraded [here](https://github.com/BLSQ/bluesquare-components/pull/121)

beygorghor commented:
my config for coda:

`THEME_PRIMARY_COLOR="#007DBC"
THEME_PRIMARY_BACKGROUND_COLOR="#eff2f5"
THEME_SECONDARY_COLOR="#085387"
APP_TITLE="Coda 2"
FAVICON_PATH="images/coda/favicon.png"
LOGO_PATH="images/coda/logo.svg"`

olethanh commented:
Not sure if the .env is the proper place for this, as it add a lot of special handling and complexity.  I think I would have preferred  to have it  in the plugin. But why not

madewulf commented:
I'm not sur how to deploy this on beanstalk. Does that mean that I have to commit the env file?

beygorghor committed:
upgrade packages and disable all maps

- lealfet to 1.9.3
- react-leaflet to 3.2.5
- comment custom Zoom Control
- return null on each map

beygorghor committed:
controls, geojsonmap

beygorghor committed:
Marker map

beygorghor committed:
planning and entities list map

beygorghor committed:
instances map + fix css issue on popup

beygorghor committed:
toggle cluster legend

beygorghor committed:
registry

beygorghor committed:
org units map :fire:

beygorghor committed:
wip calendae

beygorghor committed:
calendar map

beygorghor committed:
Calendar map

beygorghor committed:
mapcomponent

beygorghor committed:
orgunit map

beygorghor committed:
clean a bit

beygorghor committed:
small fix on org unit detail bounds calculation

beygorghor committed:
Merge branch 'main' into IA-632-leaflet

beygorghor committed:
Merge branch 'main' into IA-632-leaflet

beygorghor committed:
adding guidelines

beygorghor committed:
Write some docs

beygorghor committed:
Merge branch 'main' into IA-632-leaflet

beygorghor committed:
Merge branch 'main' into IA-632-leaflet

beygorghor committed:
Merge branch 'main' into IA-632-leaflet

beygorghor committed:
bluesquare-components

beygorghor committed:
fix edit marker

beygorghor committed:
Merge branch 'main' into IA-632-leaflet

beygorghor committed:
black

beygorghor committed:
Merge branch 'main' into IA-632-leaflet

quang-le commented on hat/assets/js/apps/Iaso/components/maps/GeoJsonMapComponent.tsx:
```diff
@@ -28,52 +34,56 @@ type Props = {
+        if (ref.current) {
+            ref.current.setUrl(currentTile.url);
+        }
+    }, [currentTile]);
```
I don't understand this part. It seems to me this effect will never run, as `ref.current` is initialized with `null` and is not set anywhere else in the component

quang-le commented on hat/assets/js/apps/Iaso/components/maps/GeoJsonMapComponent.tsx:
```diff
@@ -28,52 +34,56 @@ type Props = {
-            maxZoom: tiles.osm.maxZoom,
-            padding: boundsOptions.padding,
-        });
+    const boundsOptions: Record<string, any> = {
```
memo

quang-le commented on hat/assets/js/apps/Iaso/components/maps/tools/CustomTileLayer.tsx:
```diff
@@ -0,0 +1,29 @@
+};
+
+export const CustomTileLayer: FunctionComponent<Props> = ({ currentTile }) => {
+    const ref: any = useRef(null);
```
`RefObject<any>`

quang-le commented on hat/assets/js/apps/Iaso/components/maps/tools/CustomZoomControl.tsx:
```diff
@@ -0,0 +1,79 @@
+        const control = L.control.zoom({
+            zoomBoxTitle: formatMessage(MESSAGES['box-zoom-title']),
+            zoomInfoTitle: formatMessage(MESSAGES['info-zoom-title']),
+            fitToBoundsTitle: formatMessage(MESSAGES['fit-to-bounds']),
```
Why not `MESSAGES.boxZoomTitle` etc? I see no apparent reason not to use camelCase for the message keys

quang-le commented on hat/assets/js/apps/Iaso/components/maps/tools/CustomZoomControl.tsx:
```diff
@@ -0,0 +1,79 @@
+    }, [bounds, boundsOptions, map]);
+    const context = useLeafletContext();
+
+    L.control.zoom = opts => {
```
`opts` could be typed

quang-le commented on hat/assets/js/apps/Iaso/components/maps/tools/MapToggleCluster.tsx:
```diff
@@ -0,0 +1,52 @@
+}));
+
+type Props = {
+    isClusterActive: boolean;
+    setIsClusterActive: Dispatch<SetStateAction<boolean>>;
```
it just occurred to me that when we want to pass state+setState as props, we could just do 
```javascript
const activeClusterState = useState<boolean>()
// some code
<MapToggleCluster clusterState = {activeClusterState}/>
```

Then in `MapToggleCluster`:
```javascript
const [isClusterActive, setIsClusterActive] = clusterState
```

Not sure we gain much by doing that though, and the linter may complain that the setter should be wrapped in useEffect. 

I just don't like passing these values separately, but passing them together and pattern matching them in the child is not really better ☹️

beygorghor replied:
the isClusterActive is used in the parent component too sometimes ... i'll leave it like that for now

quang-le commented on hat/assets/js/apps/Iaso/components/maps/tools/TileSwitchComponent.js:
As far as I can see, this component is only used by `TileSwitchDialog`, so we could remove `injectIntl` and `withStyles` even if we don't convert it to arrow function without too much risk. Bonus points for TS conversion

quang-le commented on hat/assets/js/apps/Iaso/domains/instances/components/InstancesMap/useResetMapReducerOnUnmount.ts:
🍾

quang-le commented on hat/assets/js/apps/Iaso/domains/orgUnits/components/orgUnitMap/OrgUnitMap/useRedux.ts:
Only one left...

quang-le commented on hat/assets/js/apps/Iaso/components/maps/MarkerMapComponent.tsx:
```diff
@@ -41,61 +34,49 @@ export const MarkerMap: FunctionComponent<Props> = ({
     const classes: Record<string, string> = useStyles();
 
-    const boundsOptions = { padding: [500, 500] };
+    const boundsOptions: Record<string, any> = {
```
`type BoundOptions = Record<string, [number,number] | number>`

quang-le commented on hat/assets/js/apps/Iaso/domains/entities/components/ListMap.tsx:
```diff
@@ -67,63 +65,47 @@ export const ListMap: FunctionComponent<Props> = ({
     const classes: Record<string, string> = useStyles();
     const theme = useTheme();
 
+    const bounds = useMemo(
```
const bounds : Bounds | undefined ?

quang-le commented on hat/assets/js/apps/Iaso/domains/orgUnits/components/OrgUnitsMap.tsx:
```diff
@@ -104,15 +83,15 @@ const getFullOrgUnits = orgUnits => {
     const shapeBounds =
         orgUnits.shapes.length > 0 ? getShapesBounds(orgUnits.shapes) : null;
-    let bounds = null;
+    let bounds: Bounds | undefined;
```
nitpick, but style wise,  I'd prefer:
```javascript
if (locationsBounds && shapeBounds) {
    return locationsBounds.extend(shapeBounds);
    } else if (locationsBounds) {
       return locationsBounds;
    } else if (shapeBounds) {
       return shapeBounds;
    }
return undefined
```

quang-le reviewed (changes requested):
THE FIT TO BOUNDS NIGHTMARE IS FINALLY OVER!

Nothing really deep to say about the code.

- There's a z-index bug with the tooltips:

![Screenshot 2023-05-31 at 14 46 39](https://github.com/BLSQ/iaso/assets/38907762/6ef79003-3668-4515-9b15-9e2d5d5211b3)

- In registry: it's impossible to close the pop-up

https://github.com/BLSQ/iaso/assets/38907762/9027332a-c4e0-4451-bf54-8e241ec0a176

- In org units details map, the fit to bounds seems a bit off:


https://github.com/BLSQ/iaso/assets/38907762/ddcb608e-3744-4e4f-96b9-68f03f454886

beygorghor committed:
code review

beygorghor committed:
code review

beygorghor commented:
@Son, regarding the last video, can you check that the parent has a shape and is very small ? i have the same problem on my side with this org unit

quang-le commented:
> @son, regarding the last video, can you check that the parent has a shape and is very small ? i have the same problem on my side with this org unit

Yes, indeed. I tried an org unit with no parent and the fit to bounds was ok

quang-le reviewed (approved):
LET'S MERGE!!
